data/src/main/java/com/linkedin/data/schema/annotation/SchemaAnnotationHandler.java

```diff
@@ -144,6 +149,93 @@ default SchemaVisitor getVisitor()
     return new PathSpecBasedSchemaAnnotationVisitor(this);
   }
 
+  /**
+   * Check annotations changes are compatible or not
+   *
+   * @param olderResolvedProperties the older resolvedProperties
+   * @param newerResolvedProperties the newer resolvedProperties
+   * @param olderContext the older annotationCheck context
+   * @param newerContext the newer annotationCheck context
+   * @return AnnotationCompatibilityResult
+   */
+  default AnnotationCompatibilityResult annotationCompatibilityCheck(Map<String,Object> olderResolvedProperties, Map<String, Object> newerResolvedProperties,
+      AnnotationCheckContext olderContext, AnnotationCheckContext newerContext)
+  {
+    return new AnnotationCompatibilityResult();
+  }
+
+  /**
+   * AnnotationCheckContext which contains metadata information:
+   * dataSchema, schemaField, unionMember and pathToSchema.
+   */
+  class AnnotationCheckContext
+  {
+    DataSchema _dataSchema;
+    RecordDataSchema.Field _schemaField;
+    UnionDataSchema.Member _unionMember;
+    String[] _pathToSchema;
+
+    public DataSchema getDataSchema()
+    {
+      return _dataSchema;
+    }
+    public void setDataSchema(DataSchema recordDataSchema)
+    {
+      _dataSchema = recordDataSchema;
+    }
+
+    public RecordDataSchema.Field getSchemaField()
+    {
+      return _schemaField;
+    }
+    public void setSchemaField(RecordDataSchema.Field schemaField)
+    {
+      _schemaField = schemaField;
+    }
+
+    public UnionDataSchema.Member getUnionMember()
+    {
+     return _unionMember;
+    }
+    public void setUnionMember(UnionDataSchema.Member unionMember)
+    {
+      _unionMember = unionMember;
+    }
+    public String[] getPathToSchema()
+    {
+      return _pathToSchema;
+    }
+    public void setPathToSchema(ArrayDeque<String> pathToSchema)
+    {
+      _pathToSchema = pathToSchema.toArray(new String[0]);
+    }
+  }
+
+  /**
+   * AnnotationCompatibilityResult
+   * it contains a list of {@link CompatibilityMessage}
+   * CompatibilityMessage describes the change is compatible or not
+   */
+  class AnnotationCompatibilityResult implements CompatibilityResult
```

Comment:
Similar to my other comment, maybe we can add a flag to tell whether this check has been conducted or not.

Reply:
I will add a flag in this class.